File: bitmarket-common/src/main/java/com/bitwise/bitmarket/common/currency/FiatAmount.java
package com.bitwise.bitmarket.common.currency;

import java.math.BigDecimal;
import java.util.Currency;

public class FiatAmount {

    private final BigDecimal amount;
    private final Currency currency;

    public FiatAmount(BigDecimal amount, Currency currency) {
        this.amount = amount;
        this.currency = currency;
    }

    public BigDecimal getAmount() {
        return this.amount;
    }

    public Currency getCurrency() {
        return this.currency;
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result
                + (this.amount == null ? 0 : this.amount.hashCode());
        result = prime * result
                + (this.currency == null ? 0 : this.currency.hashCode());
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (this.getClass() != obj.getClass()) {
            return false;
        }
        FiatAmount other = (FiatAmount) obj;
        if (this.amount == null) {
            if (other.amount != null) {
                return false;
            }
        } else if (!this.amount.equals(other.amount)) {
            return false;
        }
        if (this.currency == null) {
            if (other.currency != null) {
                return false;
            }
        } else if (!this.currency.equals(other.currency)) {
            return false;
        }
        return true;
    }
}

File: coinffeine-common/src/main/scala/com/coinffeine/common/currency/FiatAmount.scala
package com.coinffeine.common.currency

import java.util.Currency
import scala.math.Ordering
import com.coinffeine.common.{CurrencyAmount, FiatCurrency}

@Deprecated
case class FiatAmount(amount: BigDecimal, currency: Currency) {

  /** Alternative constructor for Java code */
  def this(amount: java.math.BigDecimal, currency: Currency) =
    this(new BigDecimal(amount), currency)

  def +(that: FiatAmount): FiatAmount = {
    FiatAmount.requireCompatible(this, that)
    FiatAmount(this.amount + that.amount, this.currency)
  }

  def -(that: FiatAmount): FiatAmount = {
    FiatAmount.requireCompatible(this, that)
    FiatAmount(this.amount - that.amount, this.currency)
  }

  def unary_- : FiatAmount = FiatAmount(-amount, currency)

  def / (divisor: BigDecimal): FiatAmount = copy(amount = amount / divisor)

  override def toString = "%s %s".format(amount.underlying().toPlainString, currency)

  def toCurrencyAmount[C <: FiatCurrency]: CurrencyAmount[C] =
    FiatCurrency(currency).amount(amount).asInstanceOf[CurrencyAmount[C]]
}

@Deprecated
object FiatAmount extends Ordering[FiatAmount] {

  override def compare(x: FiatAmount, y: FiatAmount): Int = {
    requireCompatible(x, y)
    x.amount.compare(y.amount)
  }

  private def requireCompatible(x: FiatAmount, y: FiatAmount): Unit = {
    require(x.currency == y.currency)
  }
}
